pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,41 +224,234 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+			klog.InfoS("The Gateway has been updated")
+			return true
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.queue.Forget(key)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", key)
 	}
 	return true
 }
 
-func (c *MCRouteController) syncMCFlows() error {
+// reconcileWireGuard reconcile WireGuard configurations in following way:
+//  1. If the key is a Gateway dummy key, it means that we should check current Node and initialize/remove corresponding
+//     WireGuard interface on the Gateway Node.
+//  2. If the key is not a Gateway dummy key, it means that a new ClusterInfoImport event comes. The controller should
+//     update WireGuard peers on the Gateway Node.
+//
+// Note: MCDefaultRouteController runs only one worker to process Gateway and ClusterInfoImport. So we do not need
+// any synchronization mechanism.
+func (c *MCDefaultRouteController) reconcileWireGuard(key string) error {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		klog.ErrorS(err, "Failed to get active Gateway")
+		return err
+	}
+
+	isMyselfGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+	isGatewayItem := isGatewayKey(key)
+	if isGatewayItem {
+		// Controller will try to remove potential WireGuard interface on the host
+		// before WireGuard initialization or Gateway is changed or deleted.
+		if c.wireGuardClient != nil && (!isMyselfGateway || !c.wireGuardInitialized) {
+			if err := c.cleanUpWireGuard(); err != nil {
+				return err
+			}
+			c.wireGuardInitialized = false
+		}
+		if !isMyselfGateway {
+			klog.InfoS("Current Node is not Gateway Node, skip WireGuard reconcile")
+			return nil
+		}
+		if !c.wireGuardInitialized {
+			var initErr error
+			defer func() {
+				if initErr != nil {
+					klog.InfoS("Failed to initialize WireGuard, deleting interface and route")
+					if deleteErr := c.cleanUpWireGuard(); deleteErr != nil {
+						klog.ErrorS(deleteErr, "Failed to clean up WireGuard stale resources")
+					}
+				}
+			}()
+			if initErr = c.initializeWireGuard(); initErr != nil {
+				klog.ErrorS(initErr, "Failed to initialize WireGuard interface")
+				return initErr
+			}
+
+			if initErr = c.updateWireGuardPeers(); initErr != nil {
+				klog.ErrorS(initErr, "Failed to update WireGuard peers")
+				return initErr
+			}
+			c.wireGuardInitialized = true
+		}
+	} else {
+		if !isMyselfGateway {
+			klog.InfoS("Gateway does not exist or this Node is not Gateway, skip reconcile WireGuard")
+			return nil
+		}
+
+		// If the WireGuard has not been initialized. Some necessary configuration may not
+		// be ready. In this situation, we need to requeue the ClusterInfoImport again.
+		if !c.wireGuardInitialized {
+			return fmt.Errorf("WireGuard has not been initialized, requeue ClusterInfoImport")
```

Comment:
I still feel better to handle it in `processNextWorkItem()` to skip `syncMCFlows` when `!c.wireGuardInitialized`, and just return nil here. Note, `updateSingleWireGuardPeer` itself does not require a requeue, as `updateWireGuardPeers` is called after WG init.

Reply:
Agreed. Handle the check in `processNextWorkItem()` will avoid error log spam.  I tried and found that we need to check key type, which I think is duplicate(We will check it again in `reconcileWG`. Maybe we can pass the result to  `reconcileWireGuard` to avoid duplicate check?

Reply:
I moved `isMyselfGateway` and `isGatewayKey` to `processNextWorkItem` now. So `reconcileWireGuard` function only handles events when the current Node is a Gateway Node.